Update compiler to 3.2

fbiville opened the pull request:
@​Context field types can now be categorized in three ways:

 1. supported: it is legal for any procedure|function. It can be:
    - Log
    - TerminationGuard
    - GraphDatabaseService
    - SecurityContext
    - ProcedureTransaction
 2. restricted: it can be used if and only if the procedure
 |function is declared in `dbms.security.procedures.unrestricted`
 setting. The type can be:
    - GraphDatabaseAPI
    - KernelTransaction
    - DependencyResolver
    - UserManager
    - EnterpriseAuthManager
    - SecurityLog
 3. unsupported: it is illegal (any other type, basically).

The processor cannot know whether a procedure|function is
unrestricted at the time of compilation and will, by default, emit
a warning if such a type is encountered. The existing processor
option `-AIgnoreContextWarnings` can be used to mute this.

From now on, if a type is unsupported, a compilation error will be
raised.

fbiville committed:
Update component injection validation rules

@​Context field types can now be categorized in three ways:

 1. supported: it is legal for any procedure|function. It can be:
    - Log
    - TerminationGuard
    - GraphDatabaseService
    - SecurityContext
    - ProcedureTransaction
 2. restricted: it can be used if and only if the procedure
 |function is declared in `dbms.security.procedures.unrestricted`
 setting. The type can be:
    - GraphDatabaseAPI
    - KernelTransaction
    - DependencyResolver
    - UserManager
    - EnterpriseAuthManager
    - SecurityLog
 3. unsupported: it is illegal (any other type, basically).

The processor cannot know whether a procedure|function is
unrestricted at the time of compilation and will, by default, emit
a warning if such a type is encountered. The existing processor
option `-AIgnoreContextWarnings` can be used to mute this.

From now on, if a type is unsupported, a compilation error will be
raised.

fbiville commented on community/procedure-compiler/processor/src/main/java/org/neo4j/tooling/procedure/visitors/ContextFieldVisitor.java:
```diff
@@ -19,45 +19,56 @@
+            Arrays.asList( GraphDatabaseAPI.class.getName(), KernelTransaction.class.getName(),
+                    DependencyResolver.class.getName(), UserManager.class.getName(),
+                    // the following classes are not in the compiler classpath
+                    "org.neo4j.kernel.enterprise.api.security.EnterpriseAuthManager",
```
Open question: should I pull `neo4j-security-enterprise` and `neo4j-enterprise-kernel` in provided scope instead of hardcoding the two types like this?
I think it's a bad idea, though, as the compiler would only work against procedure for Neo4j Enterprise.

fbiville replied:
Other question: for now, I assumed that these new types can both be used for procedures AND functions: is that correct? 

If not, I'll need to review this visitor to carry around the original annotated method element around (that's what the second type parameter is for in `ElementVisitor` interface).

chrisvest commented on community/procedure-compiler/processor/pom.xml:
```diff
@@ -74,6 +73,18 @@
+            <artifactId>neo4j-enterprise-kernel</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
```
Community modules should not have dependencies on anything in enterprise. Tests that needs an enterprise runtime environment can be moved to enterprise. If you need to change runtime behaviour, there are usually a way with `Service.load` or similar.

fbiville replied:
OK, I'll create some specific tests in enterprise then.

fbiville replied:
Create a new module in enterprise called `procedure-compiler-enterprise-tests` and moved some tests there. Please tell me if I'm missing some information in that new POM.

chrisvest commented:
Build failure is unrelated; another PR made an API breaking change.

fbiville commented:
@chrisvest PR is now rebased. Any pending issue I should fix or is this ready to go?

OliviaYtterbrink commented on community/procedure-compiler/processor/src/main/java/org/neo4j/tooling/procedure/visitors/ContextFieldVisitor.java:
```diff
@@ -69,36 +81,55 @@ private static String types( Set<Class<?>> supportedTypes )
+                return Stream.empty();
+            }
+
+            return Stream.of( new ContextFieldWarning( field, "@%s usage warning: found restricted type <%s> on %s.\n" +
```
This should be changed to `@%s usage warning: found unsupported restricted type <%s> on %s.` instead, or something similar, since these APIs aren't official supported APIs and are subject to change at any time.

OliviaYtterbrink commented on community/procedure-compiler/processor/src/main/java/org/neo4j/tooling/procedure/visitors/ContextFieldVisitor.java:
```diff
@@ -69,36 +81,55 @@ private static String types( Set<Class<?>> supportedTypes )
 
-        return Stream.empty();
+        return Stream.of( new ContextFieldError( field,
+                "@%s usage error: found unsupported type <%s> on field %s, expected one of: %s",
```
And this should be changed from `unsupported type` to `unknown type`

fbiville commented:
@OliviaYtterbrink fixed!

chrisvest commented:
The Tools Compile build failure is unrelated I think.

fbiville commented:
There are some issues in `org.neo4j.index.internal.gbptree.CrashGenCleaner` and unknown enum `org.neo4j.index.internal.gbptree.TreeNode.BYTE_POS_NEWGEN`.

chrisvest commented:
Should be fixed by #9047

fbiville commented:
Cool, PR now rebased.

fbiville committed:
Merge @​PerformWrites check into procedure processor

fbiville committed:
Centralize custom name extraction

fbiville committed:
Centralize field validation

Field validation was unintentionally absent in user-defined
function processing. This commit finally introduces it, by reusing
the existing procedure field validation logic.

fbiville committed:
Rename stored procedure to procedure if applicable

fbiville committed:
Centralize main processing logic

fbiville committed:
Update documented rules

fbiville committed:
Add support for aggregation functions

fbiville commented:
I'm getting close to completing support of user aggregation functions!
This PR is now a WIP and not ready for merge though.

fbiville commented:
User aggregation function support is complete! :champagne: 
Feel free to review 🔍  and comment 🔈 

The only missing parts (probably another PR):

 - [ ] update of the checked rules in procedure-compiler README file
 - [ ] generation of warnings for procedure/(aggregation) function deprecated without `@Deprecated` (not sure any processor can detect the presence of java.lang annotations, but I'll try)

fbiville commented:
Hello, I have been away for 3 weeks and see there has been no recent feedback on this.
When could this be merged? Can it make it before 3.2 goes GA?

I can make myself available for live discussion if needed.

fbiville commented:
Hi, do you have any feedback on this? Do you want me to rebase it? It would be a pity if this PR does not make it at all into Neo4j, this is a blocker for other improvements.

OliviaYtterbrink commented on community/procedure-compiler/processor/src/test/java/org/neo4j/tooling/procedure/visitors/examples/RestrictedContextTypes.java:
```diff
@@ -0,0 +1,66 @@
+    @Context
+    public ProcedureTransaction procedureTransaction;
+
+    // BELOW ARE RESTRICTED TYPES
```
I think this comment should properly reflect the fact that the classes that follow aren't public APIs and thus can be changed even in patch releases, so maybe something like:
`// BELOW ARE RESTRICTED TYPES, THESE ARE UNSUPPORTED AND SUBJECT TO CHANGE`

fbiville replied:
👍 will fix this week

fbiville replied:
Fixed :)

fbiville replied:
Fixed :)

OliviaYtterbrink reviewed (changes requested):
Sorry for not responding sooner, this looks really good, so we should get it merged asap.
I just have one minor change that I would like fixed as well.

Mats-SX commented on community/procedure-compiler/integration-tests/pom.xml:
```diff
@@ -1,4 +1,5 @@
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://maven.apache.org/POM/4.0.0"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Why is the order of these fields changed in the `pom.xml` files? Some auto-format in your IDE?

fbiville replied:
good catch, not sure where it comes from, let me fix that this week

fbiville commented:
Shall I rebase as well?

OliviaYtterbrink commented:
Yes, if you would rebase it that would be great.

fbiville committed:
Reformat POM

fbiville committed:
Document unsupported restricted types in tests

fbiville commented:
Rebase done.

OliviaYtterbrink commented on enterprise/procedure-compiler-enterprise-tests/pom.xml:
```diff
@@ -0,0 +1,89 @@
+    <parent>
+        <groupId>org.neo4j</groupId>
+        <artifactId>parent</artifactId>
+        <version>3.2.0-SNAPSHOT</version>
```
Missed these two, should be `3.2.2-SNAPSHOT`

fbiville replied:
oops, thank you for catching that, will fix it today (I shamefully did not build locally before pushing the rebase)

fbiville replied:
Busy week, totally forgot to push changes. Now done :)

fbiville committed:
Clean up and fix procedure compiler POM files

fbiville committed:
Ignore failing test

This test fails because of Javac that does not send the deferred
diagnostic to compile-testing, despite the call to print it as an
error.

fbiville commented:
I had to ignore a test because of https://github.com/google/compile-testing/issues/124.

fbiville commented:
Thank you!!
